A72-open-telemetry-tracing.md

```diff
@@ -0,0 +1,643 @@
+  with name "Inbound message read" and the following attributes:
+  * key `message.event.type` with string value "RECEIVED".
+  * key `message.message.id` with integer value of the seq no.
+  * key `message.event.size.uncompressed` with integer value
+    of uncompressed message size.
+
+## Migrate from OpenCensus to OpenTelemetry
+
+### gRPC OpenCensus API
+The existing gRPC OpenCensus tracing APIs in grpc-census plugin are different between
+languages, e.g. in grpc-java it is zero-configuration: as long as the grpc-census
+dependency exists in the classpath, the traces are automatically generated.
+In C++, there is an API for users to call to enable tracing. In Go, it is exposed 
+via stream tracers. The OpenTelemetry API will coexist with the OpenCensus API.
+We keep the grpc-census plugin to allow users who already depend on grpc-census to
+continue using it for newer grpc versions.
+
+gRPC users depending on the grpc-census plugin have non-trivial migration paths
+to OpenTelemetry. Consider the following use cases:
+1. Compatibility between a gRPC client and server as two distributed components,
+where during migration one will use OpenCensus and the other will use OpenTelemetry. 
+2. Migrate an application binary where both OpenCensus and OpenTelemetry maybe exist
+   in the dependency tree. This can be the application’s own tracing code, or gRPC
+   OpenCensus, or other dependencies that involve OpenCensus and/or OpenTelemetry.
+
+Here are the suggested solutions for both use cases. 
+
+### Migrate to OpenTelemetry: Cross-process Networking Concerns
+When users first introduce gRPC OpenTelemetry, for the time window when the
+gRPC client and server have mixed plugins of OpenTelemetry and OpenCensus, 
+spanContext can not directly propagate due to different header name and wire format. 
+To tackle this, gRPC will expose a custom `GrpcTraceBinPropagator` 
+that implements `TextMapPropagator`. This will allow gRPC to keep using `grpc-trace-bin` 
+header for context propagation and also support other propagators.
+When using `grpc-trace-bin` the OpenCensus spanContext and OpenTelemetry spanContext 
+are identical, therefore a gRPC OpenCensus client can speak with a gRPC OpenTelemetry 
+server and vice versa. It is encouraged to use `GrpcTraceBinPropagator` for the migration. 
+Using the same header greatly simplifies rollout. However, there is a caveat about 
+`GrpcTraceBinPropagator`:
+
+Currently, OpenTelemetry propagator API only supports `TextMapPropagator`,
+that is to send string key/value pairs between the client and server, which is
+different from the binary header that gRPC currently uses. The future roadmap
+to support binary propagators at OpenTelemetry is unclear. So, gRPC will use
+propagator API in TextMap format with an optimization path (Go and Java) to work
+around the lack of binary propagator API to support `grpc-trace-bin`. In fact,
+TextMap propagator does not show visible performance impact for C++, which is
```

Comment:
I think we should be a little clearer here.  Specifically, I belive that what we actually found is that using the W3C text format was performant enough that it's a viable alternative to our existing binary format.  I don't think we actually checked the performance of the slow path for the binary header, and I wouldn't expect the performance to be very good for that.  So what we really decided here was that we'd live with this bad performance for now just for the migration cases, and if that posed a performance problem for someone, we can consider implementing an alternative.

Reply:
@veblush I think Esun did do a microbenchmark that showed that doing millions of encoding and decoding was actually not that bad.